File: C#/StreamView/StreamView/View/StreamViewPanel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using StreamView.Presenter;

namespace StreamView.View
{
    public partial class StreamViewPanel : UserControl, IStreamView
    {
        private IStreamViewPresenter presenter;

        #region Constructors

        /// <summary>
        /// Constructor
        /// </summary>
        public StreamViewPanel()
            : this( 128 )
        {
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="blockSize"></param>
        public StreamViewPanel( uint blockSize )
        {
            InitializeComponent();

            presenter = new StreamViewPresenter( blockSize, null, new StreamViewPresenter.UpdateView( updateView ) );

            navigatorPanel.PrevButtonPressed += new EventHandler( prevPressed );
            navigatorPanel.NextButtonPressed += new EventHandler( nextPressed );
            navigatorPanel.GotoButtonPressed += new EventHandler( gotoPressed );

            updateNavigatorControls();
        }

        #endregion

        #region Methods

        /// <summary>
        /// Updates the navigator panel's controls and labels with current
        /// data from the Presenter.
        /// </summary>
        private void updateNavigatorControls()
        {
            navigatorPanel.PrevEnabled = presenter.HasPrev;
            navigatorPanel.NextEnabled = presenter.HasNext;
            navigatorPanel.GotoEnabled = true;
            navigatorPanel.LabelText =
                "File: " + presenter.CurrentFileName + " | " +
                "Size: " + presenter.CurrentFileSize + " bytes | " +
                "Showing block " + presenter.BlockIndex + " of " + presenter.BlockCount;
            navigatorPanel.GotoText = "" + presenter.BlockIndex;
            navigatorPanel.Progress = (int)( ( ( presenter.BlockIndex * 1.0 ) / ( presenter.BlockCount == 0 ? 1 : presenter.BlockCount ) ) * 100.0 );
        }

        /// <summary>
        /// Updates the byte stream view with the current block of data.
        /// </summary>
        private void updateByteStreamView()
        {
            byteStreamView.setBytes( presenter.CurrentBlock,
                ( ( presenter.BlockIndex - 1 ) * presenter.BlockSize ) );
        }

        /// <summary>
        /// Updates the view.
        /// </summary>
        private void updateView()
        {
            updateNavigatorControls();
            updateByteStreamView();
        }

        #endregion

        #region IStreamView Methods

        /// <summary cref="IStreamView.nextPressed">
        /// <see cref="IStreamView.nextPressed"/>
        /// </summary>
        public void nextPressed( object sender, EventArgs e )
        {
            presenter.gotoNextBlock();
        }

        /// <summary cref="IStreamView.prevPressed">
        /// <see cref="IStreamView.prevPressed"/>
        /// </summary>
        public void prevPressed( object sender, EventArgs e )
        {
            presenter.gotoPrevBlock();
        }

        /// <summary cref="IStreamView.gotoPressed">
        /// <see cref="IStreamView.gotoPressed"/>
        /// </summary>
        public void gotoPressed( object sender, EventArgs e )
        {
            uint block = 0;
            try
            {
                block = UInt32.Parse( navigatorPanel.GotoText );
            }
            catch( Exception )
            {
                MessageBox.Show( "Invalid entry: " + navigatorPanel.GotoText, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error );
                return;
            }
            presenter.gotoBlock( block );
        }

        /// <summary cref="IStreamView.highlightSelection">
        /// <see cref="IStreamView.highlightSelection"/>
        /// </summary>
        public void highlightSelection( int bitOffset, int bitLength )
        {
            byteStreamView.highlightSelection( bitOffset, bitLength );
        }

        /// <summary cref="IStreamView.openFile">
        /// <see cref="IStreamView.openFile"/>
        /// </summary>
        /// <param name="filename"></param>
        public void openFile( string filename )
        {
            presenter.setInputFile( filename );
            presenter.gotoBlock( 1 );
        }

        #endregion
    }
}

File: Java/StreamView/src/view/StreamViewPanel.java
package view;

import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;
import java.io.IOException;

import javax.swing.JPanel;

import presenter.IStreamViewPresenter;
import presenter.IViewUpdater;
import presenter.StreamViewPresenter;

/*******************************************************************************
 * Simple {@link IStreamView}.
 ******************************************************************************/
public class StreamViewPanel extends JPanel implements IStreamView
{
    private IStreamViewPresenter presenter;
    private IViewUpdater viewUpdater;
    private NavigatorPanel navigator;
    private ByteStreamView byteStreamView;

    /***************************************************************************
     * Constructor
     * 
     * @throws Exception
     **************************************************************************/
    public StreamViewPanel() throws Exception
    {
        this( 128 );
    }

    /***************************************************************************
     * Constructor
     * 
     * @param blockSize
     * @throws Exception
     **************************************************************************/
    public StreamViewPanel( int blockSize ) throws Exception
    {
        viewUpdater = new ViewUpdater();
        presenter = new StreamViewPresenter( blockSize, null, viewUpdater );
        navigator = new NavigatorPanel( new PrevActionListener(),
                new NextActionListener(), new GotoActionListener() );
        byteStreamView = new ByteStreamView();

        setupPanel();

        updateNavigatorControls();
    }

    /***************************************************************************
     * Sets up this panel.
     **************************************************************************/
    private void setupPanel()
    {
        setLayout( new GridBagLayout() );
        add( navigator, new GridBagConstraints( 0, 0, 1, 1, 1.0, 0.01,
                GridBagConstraints.CENTER, GridBagConstraints.HORIZONTAL,
                new Insets( 3, 3, 3, 3 ), 0, 0 ) );
        add( byteStreamView, new GridBagConstraints( 0, 1, 1, 1, 1.0, 0.99,
                GridBagConstraints.CENTER, GridBagConstraints.BOTH, new Insets(
                        3, 3, 3, 3 ), 0, 0 ) );
    }

    /***************************************************************************
     * Updates all navigator controls.
     **************************************************************************/
    private void updateNavigatorControls()
    {
        navigator.setPrevEnabled( presenter.hasPrev() );
        navigator.setNextEnabled( presenter.hasNext() );
        navigator.setGotoEnabled( true );
        navigator.setLabelText( "File: " + presenter.getCurrentFilename()
                + " | " + "Size: " + presenter.getCurrentFilesize()
                + " bytes | " + "Showing block " + presenter.getBlockIndex()
                + " of " + presenter.getBlockCount() );
        navigator.setGotoText( "" + presenter.getBlockIndex() );
        navigator
                .setProgress( (int)( ( ( presenter.getBlockIndex() * 1.0 ) / ( presenter
                        .getBlockCount() == 0 ? 1 : presenter.getBlockCount() ) ) * 100.0 ) );
    }

    /***************************************************************************
     * Updates the displayed byte stream.
     **************************************************************************/
    private void updateByteStreamView()
    {
        byteStreamView
                .setBytes( presenter.getCurrentBlock(), ( ( presenter
                        .getBlockIndex() - 1 ) * presenter.getBlockSize() ) );
    }

    /***************************************************************************
     * Updates the display.
     **************************************************************************/
    private void updateView()
    {
        updateNavigatorControls();
        updateByteStreamView();
    }

    /*
     * (non-Javadoc)
     * 
     * @see view.IStreamView#nextPressed()
     */
    @Override
    public void nextPressed()
    {
        try
        {
            presenter.gotoNextBlock();
        } catch( IOException e )
        {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see view.IStreamView#prevPressed()
     */
    @Override
    public void prevPressed()
    {
        try
        {
            presenter.gotoPrevBlock();
        } catch( IOException e )
        {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see view.IStreamView#gotoPressed()
     */
    @Override
    public void gotoPressed()
    {
        int block = 0;
        try
        {
            block = Integer.parseInt( navigator.getGotoText() );
            presenter.gotoBlock( block );
        } catch( IOException e )
        {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see view.IStreamView#highlightSelection(int, int)
     */
    @Override
    public void highlightSelection( int bitOffset, int bitLength )
    {
        byteStreamView.highlightSelection( bitOffset, bitLength );
    }

    /*
     * (non-Javadoc)
     * 
     * @see view.IStreamView#openFile(java.io.File)
     */
    @Override
    public void openFile( File f )
    {
        try
        {
            presenter.setInputFile( f );
            presenter.gotoBlock( 1 );
        } catch( Exception e )
        {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    private class ViewUpdater implements IViewUpdater
    {
        @Override
        public void updateView()
        {
            StreamViewPanel.this.updateView();
        }
    }

    private class PrevActionListener implements ActionListener
    {
        @Override
        public void actionPerformed( ActionEvent arg0 )
        {
            StreamViewPanel.this.prevPressed();
        }
    }

    private class NextActionListener implements ActionListener
    {
        @Override
        public void actionPerformed( ActionEvent arg0 )
        {
            StreamViewPanel.this.nextPressed();
        }
    }

    private class GotoActionListener implements ActionListener
    {
        @Override
        public void actionPerformed( ActionEvent arg0 )
        {
            StreamViewPanel.this.gotoPressed();
        }
    }
}
